### ISSUE TITLE
Horovod 0.18.x fails to install, failing to run `install` or `bdist_wheel`

### ISSUE THREAD

**Issue by shailensobhee** [horovod/horovod#1506]
**Environment:**
1. Framework: Tensorflow 
2. Framework version: 1.14.0
3. Horovod version: 0.18.2 (issue happens with 0.18.0, 0.18.1, 0.18.2)
4. MPI version: Intel MPI 2019 U5- but that's not important as the issue is at the install level
5. CUDA version:
6. NCCL version:
7. Python version: 3.7.4
8. OS and version: CentOS 8
9. GCC version: (version 8.2.1 20180905 (Red Hat 8.2.1-3))

**Bug report:**
The error below happens only with 0.18.x (0.18.0, 0.18.1, 0.18.2, as per test. 0.18.2 is the latest as of today). _I do not get this error with horovod version **0.16.4**_. 

The first error, i.e failing to run `install` happens with Tensorflow obtained from PyPI (i.e `pip install tensorflow==1.14.0`) and failing to run `bdist_wheel` with Intel-Tensorflow obtained from the anaconda channel (`conda install -y tensorflow==1.14.0`).

```
env HOROVOD_WITHOUT_MXNET=1 HOROVOD_WITHOUT_PYTORCH=1 pip install --no-cache-dir horovod==0.18.2
Collecting horovod==0.18.2
  Downloading https://files.pythonhosted.org/packages/2b/ae/c86526c637e427c4bac1b6eff890ae39493ddde7ff48b7bb497d0ceaef1b/horovod-0.18.2.tar.gz (2.9MB)
     |████████████████████████████████| 2.9MB 1.6MB/s 
Requirement already satisfied: cloudpickle in ./intelpython/python3.7/lib/python3.7/site-packages (from horovod==0.18.2) (1.2.2)
Requirement already satisfied: psutil in ./intelpython/python3.7/lib/python3.7/site-packages (from horovod==0.18.2) (5.6.5)
Requirement already satisfied: pyyaml in ./intelpython/python3.7/lib/python3.7/site-packages (from horovod==0.18.2) (5.1.2)
Requirement already satisfied: six in ./intelpython/python3.7/lib/python3.7/site-packages (from horovod==0.18.2) (1.12.0)
Building wheels for collected packages: horovod
  Building wheel for horovod (setup.py) ... error
  ERROR: Complete output from command /home/centos/intelpython/latest/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-3nmpxwtq/horovod/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-5srgl9qf --python-tag cp37:
  ERROR: running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/horovod
  copying horovod/__init__.py -> build/lib.linux-x86_64-3.7/horovod
  creating build/lib.linux-x86_64-3.7/horovod/keras
  copying horovod/keras/__init__.py -> build/lib.linux-x86_64-3.7/horovod/keras
  copying horovod/keras/callbacks.py -> build/lib.linux-x86_64-3.7/horovod/keras
  creating build/lib.linux-x86_64-3.7/horovod/mxnet
  copying horovod/mxnet/mpi_ops.py -> build/lib.linux-x86_64-3.7/horovod/mxnet
  copying horovod/mxnet/__init__.py -> build/lib.linux-x86_64-3.7/horovod/mxnet
  creating build/lib.linux-x86_64-3.7/horovod/torch
  copying horovod/torch/mpi_ops.py -> build/lib.linux-x86_64-3.7/horovod/torch
  copying horovod/torch/compression.py -> build/lib.linux-x86_64-3.7/horovod/torch
  copying horovod/torch/__init__.py -> build/lib.linux-x86_64-3.7/horovod/torch
  creating build/lib.linux-x86_64-3.7/horovod/_keras
  copying horovod/_keras/__init__.py -> build/lib.linux-x86_64-3.7/horovod/_keras
  copying horovod/_keras/callbacks.py -> build/lib.linux-x86_64-3.7/horovod/_keras
  creating build/lib.linux-x86_64-3.7/horovod/common
  copying horovod/common/util.py -> build/lib.linux-x86_64-3.7/horovod/common
  copying horovod/common/__init__.py -> build/lib.linux-x86_64-3.7/horovod/common
  copying horovod/common/basics.py -> build/lib.linux-x86_64-3.7/horovod/common
  creating build/lib.linux-x86_64-3.7/horovod/tensorflow
  copying horovod/tensorflow/mpi_ops.py -> build/lib.linux-x86_64-3.7/horovod/tensorflow
  copying horovod/tensorflow/util.py -> build/lib.linux-x86_64-3.7/horovod/tensorflow
  copying horovod/tensorflow/compression.py -> build/lib.linux-x86_64-3.7/horovod/tensorflow
  copying horovod/tensorflow/__init__.py -> build/lib.linux-x86_64-3.7/horovod/tensorflow
  creating build/lib.linux-x86_64-3.7/horovod/run
  copying horovod/run/run.py -> build/lib.linux-x86_64-3.7/horovod/run
  copying horovod/run/task_fn.py -> build/lib.linux-x86_64-3.7/horovod/run
  copying horovod/run/gloo_run.py -> build/lib.linux-x86_64-3.7/horovod/run
  copying horovod/run/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run
  copying horovod/run/mpi_run.py -> build/lib.linux-x86_64-3.7/horovod/run
  creating build/lib.linux-x86_64-3.7/horovod/spark
  copying horovod/spark/__init__.py -> build/lib.linux-x86_64-3.7/horovod/spark
  creating build/lib.linux-x86_64-3.7/horovod/torch/mpi_lib
  copying horovod/torch/mpi_lib/__init__.py -> build/lib.linux-x86_64-3.7/horovod/torch/mpi_lib
  creating build/lib.linux-x86_64-3.7/horovod/torch/mpi_lib_impl
  copying horovod/torch/mpi_lib_impl/__init__.py -> build/lib.linux-x86_64-3.7/horovod/torch/mpi_lib_impl
  creating build/lib.linux-x86_64-3.7/horovod/tensorflow/keras
  copying horovod/tensorflow/keras/__init__.py -> build/lib.linux-x86_64-3.7/horovod/tensorflow/keras
  copying horovod/tensorflow/keras/callbacks.py -> build/lib.linux-x86_64-3.7/horovod/tensorflow/keras
  creating build/lib.linux-x86_64-3.7/horovod/run/util
  copying horovod/run/util/cache.py -> build/lib.linux-x86_64-3.7/horovod/run/util
  copying horovod/run/util/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run/util
  copying horovod/run/util/threads.py -> build/lib.linux-x86_64-3.7/horovod/run/util
  copying horovod/run/util/network.py -> build/lib.linux-x86_64-3.7/horovod/run/util
  creating build/lib.linux-x86_64-3.7/horovod/run/rendezvous
  copying horovod/run/rendezvous/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run/rendezvous
  copying horovod/run/rendezvous/http_server.py -> build/lib.linux-x86_64-3.7/horovod/run/rendezvous
  creating build/lib.linux-x86_64-3.7/horovod/run/driver
  copying horovod/run/driver/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run/driver
  copying horovod/run/driver/driver_service.py -> build/lib.linux-x86_64-3.7/horovod/run/driver
  creating build/lib.linux-x86_64-3.7/horovod/run/common
  copying horovod/run/common/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run/common
  creating build/lib.linux-x86_64-3.7/horovod/run/task
  copying horovod/run/task/task_service.py -> build/lib.linux-x86_64-3.7/horovod/run/task
  copying horovod/run/task/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run/task
  creating build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/env.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/config_parser.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/timeout.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/safe_shell_exec.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/host_hash.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/secret.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/network.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/settings.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  copying horovod/run/common/util/codec.py -> build/lib.linux-x86_64-3.7/horovod/run/common/util
  creating build/lib.linux-x86_64-3.7/horovod/run/common/service
  copying horovod/run/common/service/task_service.py -> build/lib.linux-x86_64-3.7/horovod/run/common/service
  copying horovod/run/common/service/__init__.py -> build/lib.linux-x86_64-3.7/horovod/run/common/service
  copying horovod/run/common/service/driver_service.py -> build/lib.linux-x86_64-3.7/horovod/run/common/service
  creating build/lib.linux-x86_64-3.7/horovod/spark/driver
  copying horovod/spark/driver/mpirun_rsh.py -> build/lib.linux-x86_64-3.7/horovod/spark/driver
  copying horovod/spark/driver/__init__.py -> build/lib.linux-x86_64-3.7/horovod/spark/driver
  copying horovod/spark/driver/driver_service.py -> build/lib.linux-x86_64-3.7/horovod/spark/driver
  copying horovod/spark/driver/job_id.py -> build/lib.linux-x86_64-3.7/horovod/spark/driver
  creating build/lib.linux-x86_64-3.7/horovod/spark/task
  copying horovod/spark/task/mpirun_exec_fn.py -> build/lib.linux-x86_64-3.7/horovod/spark/task
  copying horovod/spark/task/task_service.py -> build/lib.linux-x86_64-3.7/horovod/spark/task
  copying horovod/spark/task/__init__.py -> build/lib.linux-x86_64-3.7/horovod/spark/task
  running build_ext
  gcc -pthread -B /home/centos/intelpython/latest/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -std=c++11 -fPIC -O2 -Wall -I/home/centos/intelpython/latest/include/python3.7m -c build/temp.linux-x86_64-3.7/test_compile/test_cpp_flags.cc -o build/temp.linux-x86_64-3.7/test_compile/test_cpp_flags.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  gcc -pthread -shared -B /home/centos/intelpython/latest/compiler_compat -L/home/centos/intelpython/latest/lib -Wl,-rpath=/home/centos/intelpython/latest/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-x86_64-3.7/test_compile/test_cpp_flags.o -o build/temp.linux-x86_64-3.7/test_compile/test_cpp_flags.so
  gcc -pthread -B /home/centos/intelpython/latest/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/centos/intelpython/latest/include/python3.7m -c build/temp.linux-x86_64-3.7/test_compile/test_link_flags.cc -o build/temp.linux-x86_64-3.7/test_compile/test_link_flags.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  gcc -pthread -shared -B /home/centos/intelpython/latest/compiler_compat -L/home/centos/intelpython/latest/lib -Wl,-rpath=/home/centos/intelpython/latest/lib -Wl,--no-as-needed -Wl,--sysroot=/ -Wl,--version-script=horovod.lds build/temp.linux-x86_64-3.7/test_compile/test_link_flags.o -o build/temp.linux-x86_64-3.7/test_compile/test_link_flags.so
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_qint8 = np.dtype([("qint8", np.int8, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_qint16 = np.dtype([("qint16", np.int16, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_qint32 = np.dtype([("qint32", np.int32, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    np_resource = np.dtype([("resource", np.ubyte, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_qint8 = np.dtype([("qint8", np.int8, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_qint16 = np.dtype([("qint16", np.int16, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:544: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:545: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    _np_qint32 = np.dtype([("qint32", np.int32, 1)])
  /home/centos/intelpython/latest/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
    np_resource = np.dtype([("resource", np.ubyte, 1)])
  INFO: Compiler /usr/bin/g++ (version 8.2.1 20180905 (Red Hat 8.2.1-3)) is not usable for this TensorFlow installation. Require g++ (version >=4.8.5 20150623 (Red Hat 4.8.5-39), <5).
  INFO: Unable to build TensorFlow plugin, will skip it.
  
  Traceback (most recent call last):
    File "/tmp/pip-install-3nmpxwtq/horovod/setup.py", line 1375, in build_extensions
      build_tf_extension(self, options)
    File "/tmp/pip-install-3nmpxwtq/horovod/setup.py", line 907, in build_tf_extension
      'Could not find compiler compatible with this TensorFlow installation.\n'
  distutils.errors.DistutilsPlatformError: Could not find compiler compatible with this TensorFlow installation.
  Please check the Horovod website for recommended compiler versions.
  To force a specific compiler version, set CC and CXX environment variables.
  
  error: None of TensorFlow, PyTorch, or MXNet plugins were built. See errors above.
  ----------------------------------------
  ERROR: Failed building wheel for horovod
  Running setup.py clean for horovod
Failed to build horovod

```

**Comment by tgaddair** [comment]
Hey @shailensobhee, this behavior is intended.  The error message explains what the problem is:

```
INFO: Compiler /usr/bin/g++ (version 8.2.1 20180905 (Red Hat 8.2.1-3)) is not usable for this TensorFlow installation. Require g++ (version >=4.8.5 20150623 (Red Hat 4.8.5-39), <5).
```

The reason you don't get this behavior in older versions of Horovod is because we didn't run this compiler check previously.  In older versions of Horovod, the build process would "succeed", but then could segfault at runtime due to the ABI incompatibility.

My suggestion would be to use TensorFlow 1.15 (which was built with a newer version of g++), or to install an older version of g++ and set the `CC` and `CXX` environment variables to use a specific compiler version.

**Comment by shailensobhee** [comment]
I use `conda` as my package manager and Tensorflow version `1.15.x` is not available on the `anaconda` channel. From the 1.1x line, it seems that `1.14.0` is the latest I can use. I cannot use the Tensorflow from PyPI (i.e `pip install tensorflow` is not an option for me).

Any recommendations?

**Comment by tgaddair** [comment]
Hey @shailensobhee, can you check which versions of gcc are available:

```
yum list available | grep gcc
yum list available | grep devtoolset
```

If it's available, try installing GCC 4.8.  As long as it's on your system and in your binary path, it Horovod should be able to use it to install TensorFlow.
